### ISSUE TITLE
Possible memory leak and not thread safe singleton

### ISSUE THREAD

**Issue by gestern** [newnon/avalon#5]
It looks like a simultaneously a non thread-safe singleton and aligning it with the paradigm PImpl.

https://github.com/newnon/avalon/blob/52ff1f0051520053c88e16f37cb43f023a1acac0/avalon/platform/ios/payment/payment.mm#L305-L310

**Comment by newnon** [comment]
most part of cocos2d-x not thread safe. avalon also does not thread safe 
every function must be called inside lock or something similar even "getInstance()"

**Comment by gestern** [comment]
> most part of cocos2d-x not thread safe. avalon also does not thread safe
> every function must be called inside lock or something similar even "getInstance()"

Yes, I agree, but in the cocos2d-x they are executed all the same in one of three main isolated threads: either a render or I/O or networks and they are still synchronized with each other. In any case, there is **a memory leak** even if we do not take into consideration the race in the initialization of the singleton.
